### PR TITLE
Attach iso

### PR DESCRIPTION
**Pull request by wanyufe**
*Issue #, if available:*
Feature addition: ISO attachment to VM feature.
*Description of changes:*
Allow an ISO to attach to VM before VM starts.
*Testing performed:*
Unit testing and manual testing

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: api/v1beta1/cloudstackmachine_types.go [hunk 1]
```diff
@@ -52,6 +52,10 @@ type CloudStackMachineSpec struct {
 	// +optional
 	DiskOffering CloudStackResourceDiskOffering `json:"diskOffering,omitempty"`
 
+	// CloudStack ISO attachment, ISO is attached before VM starts.
+	// +optional
+	ISOAttachment CloudStackResourceIdentifier `json:"ISOAttachment,omitempty"`
```
Comment (davidjumani) [inline api/v1beta1/cloudstackmachine_types.go:57]: I'm concerned about the standardization of the json tags. Over here it matches the struct member,s name but for sshkey it does not. Can this be standardized ?

Comment (davidjumani) [inline api/v1beta1/cloudstackmachine_types.go:57]: I'm concerned about the standardization of the json tags. Over here it matches the struct member,s name but for sshkey it does not. Can this be standardized ?

File: controllers/cloudstackmachine_controller.go [hunk 2]
```diff
@@ -199,14 +194,55 @@ func (r *CloudStackMachineReconciliationRunner) GetOrCreateVMInstance() (retRes
...
+// AttachISOIfConfigured attach ISO to vm instance
+func (r *CloudStackMachineReconciliationRunner) AttachISOIfConfigured() (retRes ctrl.Result, err error) {
+	if len(r.ReconciliationSubject.Spec.ISOAttachment.ID) == 0 && len(r.ReconciliationSubject.Spec.ISOAttachment.Name) == 0 {
+		// ISO attachment not configured, ignore
+		return ctrl.Result{}, nil
+	}
+	if len(*r.ReconciliationSubject.Spec.InstanceID) == 0 {
+		return r.RequeueWithMessage("vm instance not yet available.")
+	}
+	r.Log.Info("Start to attach ISO.")
```
Comment (maxdrib) [inline controllers/cloudstackmachine_controller.go:230]: might be helpful to include some identifiers here - for the ISOAttachment and the VM instance

**Comment by rejoshed** [comment]
@wanyufe can we close this PR? Is it still something you want to merge?

**Comment by rejoshed** [comment]
This feature may be useful in the future, but the branch and PR are too outdated. Closing.

**Pull request closed without merging**